File: results/local/2026-09-09/current-fleet-composition-v628/reproduce/package.mjs
import assert from 'node:assert/strict';
import { readFile, writeFile, copyFile, mkdir, readdir, stat } from 'node:fs/promises';
import { createHash } from 'node:crypto';
import { gzipSync, gunzipSync } from 'node:zlib';
import { dirname, resolve, join, relative } from 'node:path';
import { fileURLToPath } from 'node:url';

const kit = dirname(fileURLToPath(import.meta.url));
const root = resolve(kit, '../../..');
const output = join(root, 'results/local/2026-09-09/current-fleet-composition-v628');
const campaign = join(root, 'build/performance/seeded-current-composition-v628');
const review = join(root, 'build/performance/current-composition-review-v628');
const previous = join(root, 'build/performance/seeded-candidate-boundary-merit-v627/campaign');
const base = join(root, 'results/lambda/2026-09-09/gpu-regeneration-v799/h100-best');
const viewer = join(root, 'results/lambda/2026-09-06/visualiser');
const hash = data => createHash('sha256').update(data).digest('hex');
const read = async p => JSON.parse(await readFile(p, 'utf8'));
const write = async (name, value) => {
  const path = join(output, name);
  await mkdir(dirname(path), { recursive: true });
  await writeFile(path, typeof value === 'string' || Buffer.isBuffer(value) ? value : JSON.stringify(value, null, 2) + '\n', { flag: 'wx' });
};
const copied = [];
const copy = async (source, target) => {
  const data = await readFile(source);
  await write(target, data);
  copied.push({ path: target, source: relative(root, source).replaceAll('\\', '/'), bytes: data.length, sha256: hash(data) });
};
async function files(directory) {
  const result = [];
  for (const entry of await readdir(directory, { withFileTypes: true })) {
    const path = join(directory, entry.name);
    if (entry.isDirectory()) result.push(...await files(path));
    else if (entry.isFile()) result.push(path);
    else throw new Error(`Unexpected non-file entry: ${path}`);
  }
  return result.sort();
}
function sections(data) {
  const result = new Map();
  for (const line of data.toString('latin1').match(/[^\n]*\n|[^\n]+$/g) || []) {
    const ship = line.trim().split(/\s+/)[0];
    if (!ship) continue;
    assert(/^\d+$/.test(ship));
    result.set(ship, Buffer.concat([result.get(ship) || Buffer.alloc(0), Buffer.from(line, 'latin1')]));
  }
  return result;
}

await assert.rejects(stat(output), { code: 'ENOENT' });
await mkdir(output);
await write('.gitattributes', '* -text -whitespace\n');
const plan = await read(join(campaign, 'plan.json'));
const sourceBindings = [];
for (const [name, expected] of Object.entries(plan.source_sha256)) {
  const bytes = await readFile(join(root, name));
  assert.equal(hash(bytes), expected, name);
  sourceBindings.push({ path: name, bytes: bytes.length, sha256: expected });
}
await copy(join(campaign, 'inputs/Result.txt'), 'Result.txt');
assert.equal(hash(await readFile(join(output, 'Result.txt'))), plan.composed_result_sha256);
for (const file of ['report.json', 'checker-binding.json']) await copy(join(campaign, 'output', file), file);
for (const file of ['independent-full.json', 'independent.json', 'official-full.json', 'official.json', 'partial-report.json', 'launch-report.json', 'worker.log', 'official/ScoreData.txt']) {
  await copy(join(campaign, 'output', file), `raw/${file}`);
}
for (const file of ['prepare.py', 'run.py', 'plan.json', 'launch-marker.json']) await copy(join(campaign, file), `run/${file}`);
for (const path of await files(review)) await copy(path, `review/${relative(review, path).replaceAll('\\', '/')}`);
for (const [source, target] of [
  [join(base, 'campaign-report.json'), 'lineage/v799-campaign-report.json'],
  [join(previous, 'output/report.json'), 'lineage/v627-report.json'],
  [join(previous, 'output/fleet/checker-binding.json'), 'lineage/v627-checker-binding.json'],
  [join(previous, 'ready.json'), 'lineage/v627-ready.json'],
  [join(previous, 'common.py'), 'lineage/v627-common.py'],
  [join(previous, 'profile.json'), 'lineage/v627-profile.json'],
]) await copy(source, target);
await write('lineage/source-bindings.json', { status: 'all_original_plan_inputs_rehashed', sources: sourceBindings,
  note: 'The original execution scripts retain their workspace-relative paths. Native/source lineage is published separately in endpoint-merit-and-gpu-dual-v627; no native or verifier executables are copied here.' });
const current = sections(await readFile(join(output, 'Result.txt')));
const old = sections(await readFile(join(base, 'Result.txt')));
const replacement = sections(await readFile(join(previous, 'output/fleet/Result.txt')));
const proof = { current_result_sha256: plan.composed_result_sha256, H100_v799_Result_sha256: plan.source_result_sha256,
  local_v627_Result_sha256: plan.certified_route_result_sha256, sections: {} };
assert.equal(current.size, 23);
for (const [ship, data] of current) {
  const selected = (ship === '8' ? replacement : old).get(ship);
  assert(data.equals(selected));
  proof.sections[ship] = { bytes: data.length, composed_sha256: hash(data), selected_source: ship === '8' ? 'local_v627' : 'H100_v799',
    selected_source_sha256: hash(selected), baseline_sha256: hash(old.get(ship)) };
}
await write('lineage/retained-sections.json', proof);
const archives = [];
async function compressed(source, target) {
  const data = await readFile(source), compressedBytes = gzipSync(data, { level: 9 });
  assert(gunzipSync(compressedBytes).equals(data));
  await write(target, compressedBytes);
  archives.push({ path: target, source: relative(root, source).replaceAll('\\', '/'), compressed_bytes: compressedBytes.length,
    compressed_sha256: hash(compressedBytes), uncompressed_bytes: data.length, uncompressed_sha256: hash(data), roundtrip_exact: true });
}
for (const file of ['export.mjs', 'import.mjs', 'verify-http.mjs', 'README.md', 'export-audit.json', 'import-audit.json', 'http-audit.json', 'schema.log', 'syntax.log']) {
  await copy(join(kit, file), `viewer/${file}`);
}
for (const file of ['manifest.json', 'fleet-summary.json', 'checker-binding.json', 'compute.json']) await copy(join(kit, 'export', file), `viewer/export/${file}`);
await compressed(join(kit, 'export/trajectories.json'), 'viewer/export/trajectories.json.gz');
await compressed(join(viewer, 'data/gtoc12-current-composition-v628/fleet.json'), 'viewer/installed/fleet.json.gz');
for (const file of ['manifest.json', 'checker-binding.json', 'compute.json', '.gitattributes']) {
  await copy(join(viewer, 'data/gtoc12-current-composition-v628', file), `viewer/installed/${file}`);
}
for (const file of ['app.js', 'index.html', 'scripts/import-gtoc12.mjs', 'scripts/import-data.mjs', 'kepler.js']) await copy(join(viewer, file), `viewer/source/${file}`);
await write('archive-audit.json', { status: 'all_gzip_roundtrips_byte_exact', files: archives });
await copy(join(kit, 'verify_package.py'), 'reproduce/verify_package.py');
await copy(join(kit, 'package.mjs'), 'reproduce/package.mjs');
await copy(join(kit, 'package-README.md'), 'README.md');
await write('copy-manifest.json', { files: copied, excluded_duplicate_artifacts: [
  { path: 'output/official/Result.txt', reason: 'Exact canonical Result.txt already copied', sha256: plan.composed_result_sha256 },
  { path: 'output/official/GTOC12_Asteroids_Data.txt', reason: 'Official catalogue is an existing pinned external dataset; no duplicate', sha256: (await read(join(campaign, 'output/report.json'))).catalogue_sha256 },
  { path: 'output/official/GTOC12_Verify', reason: 'No executable binaries in this package', sha256: 'd4e4bc81129266420b27c9bde038bce9eda1960e7de9c695772fbfdb1cc82cd6' },
] });
const indexed = {};
for (const path of await files(output)) {
  const bytes = await readFile(path);
  indexed[relative(output, path).replaceAll('\\', '/')] = { bytes: bytes.length, sha256: hash(bytes) };
}
await write('index.json', { file_count: Object.keys(indexed).length, bytes: Object.values(indexed).reduce((a, x) => a + x.bytes, 0), files: indexed });
console.log(JSON.stringify({ output: relative(root, output).replaceAll('\\', '/'), file_count: Object.keys(indexed).length,
  index_sha256: hash(await readFile(join(output, 'index.json'))), bytes: Object.values(indexed).reduce((a, x) => a + x.bytes, 0) }));

File: results/local/2026-09-09/mass-budgeted-frontier-v629/package.py
"""Package exact saved v629 evidence; no numerical execution or source mutation."""
from __future__ import annotations
import hashlib
import json
from pathlib import Path
import shutil
import tarfile

KIT = Path(__file__).resolve().parent
ROOT = KIT.parents[2]
CAMPAIGN = ROOT / "build/performance/mass-budgeted-frontier-v629"
VIEW = ROOT / "build/performance/frontier-visualizer-v629"
DEST = ROOT / "results/local/2026-09-09/mass-budgeted-frontier-v629"
DATASET = ROOT / "results/lambda/2026-09-06/visualiser/data/gtoc12-frontier-v629"


def sha(path):
    return hashlib.sha256(path.read_bytes()).hexdigest()


def read(path):
    return json.loads(path.read_text())


def write(path, value):
    with path.open("x") as stream:
        json.dump(value, stream, indent=2, allow_nan=False)
        stream.write("\n")


def files_under(path):
    return sorted(p for p in path.rglob("*") if p.is_file())


def main():
    if not __debug__:
        raise RuntimeError("Assertions required")
    DEST.mkdir()
    ready = read(CAMPAIGN / "ready.json")
    for name, expected in ready["files"].items():
        assert sha(CAMPAIGN / name) == expected
    assert sha(CAMPAIGN / "ready.json") == "6a63b9506655dd1590c7c470a5a5314315ba565475658a82ea3adbd28cd9bf6b"
    assert sha(KIT / "inputs/Result.txt") == "765cb7ef97d38926317dbbf4ae8700626dffec06c48af3d6f68dae47c154f3da"
    assert read(KIT / "output/report.json")["qualified"]
    copies = {
        "Result.txt": KIT / "inputs/Result.txt", "RESULTS.md": CAMPAIGN / "RESULTS.md",
        "checker-binding.json": KIT / "output/checker-binding.json",
        "report.json": KIT / "output/report.json",
        "independent.json": KIT / "output/independent.json",
        "independent-full.json": KIT / "output/independent-full.json",
        "official.json": KIT / "output/official.json", "official-full.json": KIT / "output/official-full.json",
        "official.stdout": KIT / "output/official.stdout", "official.stderr": KIT / "output/official.stderr",
        "composition-launch-report.json": KIT / "output/launch-report.json",
        "campaign-report.json": CAMPAIGN / "output/report.json",
        "campaign-launch-report.json": CAMPAIGN / "output/launch-report.json",
        "campaign-ready.json": CAMPAIGN / "ready.json",
        "saved-node-provenance.json": VIEW / "export/saved-node-provenance.json",
        "viewer-export-audit.json": VIEW / "export-audit.json",
        "viewer-import-audit.json": VIEW / "import-audit.json",
        "viewer-http-audit.json": VIEW / "http-audit.json",
        "package.py": Path(__file__),
        "verify_package.py": KIT / "verify_package.py",
    }
    for name, source in copies.items():
        shutil.copyfile(source, DEST / name)
    (DEST / ".gitattributes").write_text("* -text -whitespace\n")
    assigned, omitted, archives = set(), [], {}

    def archive(name, paths):
        rows = {}
        with tarfile.open(DEST / name, "w:gz", compresslevel=6) as target:
            for path in sorted(set(paths)):
                relative = path.relative_to(ROOT).as_posix()
                if relative in assigned:
                    continue
                if path.name == "GTOC12_Verify" or path.name == "GTOC12_Asteroids_Data.txt" or "__pycache__" in path.parts or path.suffix == ".pyc":
                    omitted.append({"path": relative, "sha256": sha(path), "bytes": path.stat().st_size, "reason": "copied checker executable or duplicate public catalogue; caches omitted"})
                    continue
                assert path.is_file() and not path.is_symlink()
                data_sha = sha(path)
                rows[relative] = {"sha256": data_sha, "bytes": path.stat().st_size}
                target.add(path, arcname=relative, recursive=False)
                assigned.add(relative)
        archives[name] = {"sha256": sha(DEST / name), "bytes": (DEST / name).stat().st_size, "files": rows}

    archive("campaign.tar.gz", files_under(CAMPAIGN))
    archive("composition.tar.gz", files_under(KIT))
    archive("viewer.tar.gz", files_under(VIEW) + files_under(DATASET))
    prerequisite = []
    prior = ROOT / "build/performance/seeded-candidate-boundary-merit-v627/campaign"
    prerequisite.append(prior / "ready.json")
    prerequisite.extend(prior / name for name in read(prior / "ready.json")["files"])
    base = ROOT / "build/performance/refinement-queue-v623"
    prerequisite.append(base / "source-sha256.json")
    prerequisite.extend(base / "source" / name for name in read(base / "source-sha256.json"))
    for name in ready["external_files"]:
        path = (CAMPAIGN / name).resolve()
        if path.name in ("local.tar.gz", "local-files.json"):
            continue  # Existing committed v799 pool; hash pointer retained below.
        prerequisite.append(path)
    prerequisite.extend(ROOT / name for name in read(KIT / "checker-plan.json")["source_sha256"])
    archive("frozen-prerequisites.tar.gz", prerequisite)
    write(DEST / "archive-manifest.json", archives)
    write(DEST / "omitted-artifacts.json", omitted)
    write(DEST / "external-runtime.json", {
        "profile": read(CAMPAIGN / "profile.json"),
        "external_saved_pool": {name: digest for name, digest in ready["external_files"].items() if name.endswith(("local.tar.gz", "local-files.json"))},
        "catalogue_sha256": read(KIT / "output/report.json")["catalogue_sha256"],
        "bonus_sha256": read(KIT / "output/report.json")["bonus_sha256"],
        "no_runtime_binaries_in_package": True,
    })
    (DEST / "README.md").write_text((KIT / "package-README.md").read_text())
    index = {p.relative_to(DEST).as_posix(): {"sha256": sha(p), "bytes": p.stat().st_size} for p in files_under(DEST)}
    write(DEST / "index.json", {"files": index, "count": len(index), "bytes": sum(x["bytes"] for x in index.values())})
    print(json.dumps({"directory": str(DEST), "index_sha256": sha(DEST / "index.json"), "files": len(index), "bytes": sum(x["bytes"] for x in index.values()), "archive_members": sum(len(x["files"]) for x in archives.values()), "omitted": len(omitted)}))


if __name__ == "__main__":
    main()
